Adding cipher tag to the serverFromString

KaviHarjani opened the pull request:
## Scope and purpose

Fixes #12134

Customising serverFromString further to handle ciphers as well, 
Daphne extends the use case for twisted by the "-e" flag, this make 
Daphe and Twisted more flexible, helping it add a static cipher to accept from 

[Futher detail here](https://github.com/twisted/twisted/issues/12134)

KaviHarjani committed:
Adding cipher tag to the serverFromString

KaviHarjani committed:
Adding cipher tag to the serverFromString

KaviHarjani commented:
please review

adiroiban commented on NEWS.rst:
```diff
@@ -4417,6 +4417,7 @@ Twisted Core 15.0.0 (2015-01-24)
 
 Features
 --------
+ - twisted.internet.endpoints adding support for cipher in `serverFromString`(#12134)
```
This needs to be moved into a "fragment" file.

We don't do direct edit of the release notes.

See more info here https://docs.twisted.org/en/stable/development/dev-process.html#release-notes-management

KaviHarjani replied:
[#6e30f](https://github.com/twisted/twisted/pull/12137/commits/6e30f9f83d118633e7be1c489d00073549c1b285)

adiroiban commented on src/twisted/internet/endpoints.py:
```diff
@@ -1466,6 +1467,10 @@ def _parseSSL(
         **kw,
     )
+
+    if cipher:
+        if isinstance(cipher, str):
```
I guess that it's ok if we don't do type checking for str.


```suggestion
    if cipher is not None:
```

What should happen if you pass a boolean or other types ?

I don't think that we need to worry.

What we need to worry, is to document the `cipher` arument and let people know that it should be `str`

KaviHarjani replied:
[#692e540f](https://github.com/twisted/twisted/pull/12137/commits/692e540f85cac1146fb0a44a230138ca743ce5b0)

adiroiban commented on src/twisted/internet/endpoints.py:
```diff
@@ -1466,6 +1467,10 @@ def _parseSSL(
+
+    if cipher:
+        if isinstance(cipher, str):
+            cf.getContext().set_cipher_list(str.encode(cipher.replace(",", ":")))
```
I think is best to have explicit `ascii` encoding here.

KaviHarjani replied:
[#692e540f](https://github.com/twisted/twisted/pull/12137/commits/692e540f85cac1146fb0a44a230138ca743ce5b0)

adiroiban commented on src/twisted/internet/test/test_endpoints.py:
```diff
@@ -3259,6 +3259,29 @@ def test_sslNoTrailingNewlinePem(self):
+    @skipIf(skipSSL, skipSSLReason)
+    def test_sslWithCustomCipher(self):
+        """
+        Test adding custom cipher to the serverFromString function.
```
Maybe something like this.

See https://jml.io/test-docstrings/

```suggestion
        A cipher list is supported, using the OpenSSL format.
        The colon (:) from the OpenSSL format is replaced with a comman (').
```

adiroiban commented on src/twisted/internet/test/test_endpoints.py:
```diff
@@ -3259,6 +3259,29 @@ def test_sslNoTrailingNewlinePem(self):
+        )
+        self.assertIsInstance(server, endpoints.SSL4ServerEndpoint)
+        self.assertIs(server._reactor, reactor)
+        self.assertEqual(server._port, 1234)
+        self.assertEqual(server._backlog, 12)
+        self.assertEqual(server._interface, "10.0.0.1")
+        self.assertEqual(server._sslContextFactory.method, TLSv1_2_METHOD)
```
ar these assertions needed for this test?

I guess that here we should assert that the ciphers were set...

I don't know how to get the configured cipher list for an SSL context :(

-------

For my code, to test it, I do an actual handshake...

but I feel that that it's too much here.

adiroiban commented on src/twisted/internet/test/test_endpoints.py:
```diff
@@ -3259,6 +3259,29 @@ def test_sslNoTrailingNewlinePem(self):
+            endpoints.serverFromString(
+                reactor,
+                "ssl:1234:backlog=12:privateKey=%s:"
+                "certKey=%s:sslmethod=TLSv1_2_METHOD:interface=10.0.0.1:cipher=ECDHE-ECDSA-AES128-GCM-SHA256,ECDHE-ECDSA-CHACHA20-POLY1305,ECDHE-RSA-AES128-GCM-SHA256"
```
WOuld it make sense to keep the input for this test with a narrow scope?

Do we need backlog here ?
Do we need to care about sslmethod here ? 

I expect that we have separate tests for backlog or ssl method

```suggestion
                "certKey=%s:cipher=ECDHE-ECDSA-AES128-GCM-SHA256,ECDHE-RSA-AES128-GCM-SHA256"
```

adiroiban replied:
Maybe it would help to have a cipher text like

```
ALL:!ADH:@STRENGTH:+RSA:-DSA:SHA1+DES
```

I hope that this provide a good cover of the value that can be found in a cipher definition

adiroiban reviewed (changes requested):
Many thanks. The changes look good and are a good start.

Things blocking the merge:

* dedicated release notes fragment
* handling of non-string arguments
* lack of documentation

How are people expected to know about this feature ?

I was expected to see it documented in the docstring of serverFromString  

similar to what we have for other rules https://docs.twistedmatrix.com/en/stable/api/twisted.internet.endpoints.html#serverFromString


there is also this documentation page https://docs.twisted.org/en/stable/core/howto/endpoints.html#endpoint-types-included-with-twisted

and I think that it would help to document this feature here.

----------


It would be nice to have a followup for clientFromString ... but this needs to be done in a separate ticket/PR

We should focus on the serverFromString for this PR.

Thanks again

KaviHarjani committed:
isort changes

KaviHarjani commented:
please review

glyph commented on NEWS.rst:
```diff
@@ -3873,7 +3873,7 @@ Twisted Core 15.4.0 (2015-09-04)
 
 This is the last Twisted release where Python 2.6 is supported, on any
-platform. 
+platform.
```
This file should not be changed manually, please remove this change.

KaviHarjani replied:
This was autoformatted

glyph commented on src/twisted/internet/endpoints.py:
```diff
@@ -1466,6 +1470,10 @@ def _parseSSL(
     )
+
+    if cipher:
+        cipher_bytes = cipher.replace(",", ":").encode("ascii")
+        cf.getContext().set_cipher_list(cipher_bytes)
```
Twisted's coding standard is slightly different than PEP8; you should be using `camelCase` rather than `snake_case` for local variables.

```suggestion
        cipherBytes = cipher.replace(",", ":").encode("ascii")
        cf.getContext().set_cipher_list(cipherBytes)
```

glyph commented on src/twisted/internet/test/test_endpoints.py:
```diff
@@ -3259,6 +3259,27 @@ def test_sslNoTrailingNewlinePem(self):
+        self.assertIsInstance(server, endpoints.SSL4ServerEndpoint)
+        self.assertEqual(server._sslContextFactory.method, TLSv1_2_METHOD)
+        ctx = server._sslContextFactory.getContext()
+        self.assertIsInstance(ctx, ContextType)
```
This coverage warning indicates that this test always fails.

KaviHarjani replied:
@glyph thankyou, I had to reinstall and run the test for this to show, it was actually passing all the tests locally and I was unable to figure

glyph reviewed (dismissed):
Please get the tests passing and patch coverage to 100% before requesting another review so as to make efficient use of reviewer time; we can't integrate changes that don't pass our basic automated quality checks.

KaviHarjani committed:
Update src/twisted/internet/endpoints.py

Co-authored-by: Glyph <code@glyph.im>

KaviHarjani committed:
Fixing the test error

KaviHarjani committed:
Merge branch 'kaviharjani_12134' of github.com:KaviHarjani/twisted into kaviharjani_12134

KaviHarjani committed:
Test fix

KaviHarjani committed:
Removed TLS version as that is not being passed in the string anymore and defaults to the latest

KaviHarjani committed:
Undo NEWS.rst update

KaviHarjani commented:
please review

glyph commented on src/twisted/internet/endpoints.py:
```diff
@@ -1414,6 +1415,9 @@ def _parseSSL(
         constant in C{OpenSSL.SSL}.
     @type sslmethod: C{str}
 
+    @param cipher: A string containing a list of cipher values, formatted as comma-separated values.
```
What is a "cipher value"? Presumably there is a list of these magic strings somewhere. Is it OpenSSL specific? How should the user discover the possible list of options? Why might they want to use this option?

KaviHarjani replied:
@glyph yes, so if you do openssl ciphers -v 'ALL:eNULL', you can find all the cipher values 

The usecase is, I work as a developer in a bank and they want to limit cipher use, even in so basic requirement is TLS1.2 and even in there, they only want to enable a few ciphers

They are already using Daphne and Daphne is using `serverFromString` from Twisted 

They are averse of making big changes like using Twisted directly, so with this I can have them use only selected cipher for them

I was going to make this changes directly in Daphne but this could be used by a lot of people 
So doing it here made more sense to me

KaviHarjani replied:
Should i mention from where to get the -cipher value from? 
I think people wanting to use this feature would actually know what they are looking for and will specifically try finding it, like I id haha

Also thanks for `Twisted` you guys <3

glyph replied:
> The usecase is, I work as a developer in a bank and they want to limit cipher use, even in so basic requirement is TLS1.2 and even in there, they only want to enable a few ciphers

Sorry, I should have been clearer: I am asking these questions not because I don't know the answer (I do) but from the perspective of some random future developer reading about this in Twisted.  Twisted right now uses OpenSSL for TLS, but hopefully we will not always do that (I have a pretty short-term to-do item to look at pyrtls to see if we can get a more memory-safe option).

So, imagine that you are reading this as someone who knows what TLS is, and wants to do something with it, but doesn't know anything in particular about OpenSSL or how to use it.

glyph replied:
So the questions are all things that need to be somehow answered within the text of the source code, not just in a comment reply :)

KaviHarjani replied:
@glyph added it, let me know if its good enough or do I need to add more :)

glyph commented:
(Thanks a ton for getting those tests passing!)

glyph commented:
please review

glyph commented:
oops I did not mean to dismiss the review with my one stray comment

KaviHarjani committed:
Update cipher description

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

adiroiban commented on src/twisted/newsfragments/12134.feature:
```diff
@@ -0,0 +1 @@
+Added support for `cipher` in `serverFromString` function
```
The current test read more like a commit mesage.

The newsframgment, is for the public release notes.

It's important to clearly explain which API is targeted by this change, and what is this change all about.

The project convention for the release notes, it to start the message with the canonical name of the API that was changed

```suggestion
twisted.internet.endpoint.serverFromString now supports defining a cipher list. 
```

KaviHarjani replied:
[#f991e5a](https://github.com/twisted/twisted/pull/12137/commits/f991e5a4aa8b557cb19c222f876940b12cc0c0b6)

adiroiban commented on src/twisted/internet/test/test_endpoints.py:
```diff
@@ -3259,6 +3259,23 @@ def test_sslNoTrailingNewlinePem(self):
+        reactor = object()
+        server = endpoints.serverFromString(
+            reactor,
+            "ssl:1234:backlog=12:privateKey=%s:"
+            "certKey=%s:cipher=ECDHE-ECDSA-AES128-GCM-SHA256,ECDHE-RSA-AES128-GCM-SHA256,ALL,!ADH"
```
I see my previous comment was marked as "resolved" , but I don't see any info about why no change is required.

Do we need `backlog` in this test? Will the test fail if backlog is not provided ?

Is there any value in having both `ECDHE-ECDSA-AES128-GCM-SHA256` and `ECDHE-RSA-AES128-GCM-SHA256` suites in this list ?

I still think that we should look into a minimal list, that covers all the available OpenSSL options ... something like `ALL:!ADH:@STRENGTH:+RSA:-DSA:SHA1+DES`

I can see that this is valid

```
$ openssl ciphers 'ALL:!ADH:@STRENGTH:+RSA:-DSA:SHA1+DES'
```

-----------

I am note sure that this test provides good coverage.

It's important to have

For example, if I remove these lines from the branch, would this test fail ?

```
    if cipher:
        cipherBytes = cipher.replace(",", ":").encode("ascii")
        cf.getContext().set_cipher_list(cipherBytes)
```

--------

I still think that this test is valid.
I also think that the only way to check which ciphers were set, is to do a full handshake ... and this might be too much for this test.

But it's important to add a note about this limitation in the docstring of the test.

---------

What happens if I pass a string like `ssl:1234:cipher=bla,bla-bla`

I think that we need a test to check how the errors are reported.

I guess that the error here would be "SSL routines - no cipher match" ... but is best to have a test in which we describe that we don't have any custom error message and the OpenSSL low-level exception is raised.

I don't know if we want that.

I guess that it would be nice to handle the OpenSSL exception and show a better error message

KaviHarjani replied:
[#fe8f46a](https://github.com/twisted/twisted/pull/12137/commits/fe8f46a3f91f6efc12a73f7c6fab3fb494e17c89), [#53b0d41](https://github.com/twisted/twisted/pull/12137/commits/53b0d41b6a8aa4b90edc2365739031c951e0db13)

adiroiban reviewed (changes requested):
Many thanks for the updates.

Happy to see the release notes and the test passing.

I can see that my comments were marked as resolved, but I don't see any reply to my comments. Not a big deal... but it's a bit confusing, especially since some change requests were not made.

I left another comment.

------------

It's ok to have a test for the "happy path", but I think that we should also have a test for error handling.

The test for error handling should also provide a better coverage to make sure the low-level OpenSSL API is called.

Please consider adding a test for the case in which an invalid cipher list  value is used.

Thanks again.

KaviHarjani committed:
Update src/twisted/newsfragments/12134.feature

Co-authored-by: Adi Roiban <adiroiban@gmail.com>

KaviHarjani committed:
Cipher failing error message

KaviHarjani committed:
Limitations in test docstring

KaviHarjani committed:
Merge branch 'kaviharjani_12134' of github.com:KaviHarjani/twisted into kaviharjani_12134

KaviHarjani commented:
please review

@adiroiban thanks for the review, 
the tests are failing as it is unable to import OpenSSL module
Do you want me to handle the exceptions generically?


Or something like

```
except Exception as e:
    if "OpenSSL.SSL.Error" in str(e):
        raise Exception("Invalid cipher list passed")
    else:
        raise e
```

KaviHarjani committed:
Adding pyopenssl to GTK

KaviHarjani committed:
Adding pyopenssl in dependencies

KaviHarjani committed:
Using SSlError

KaviHarjani committed:
Update test.yaml

KaviHarjani committed:
Test update

KaviHarjani committed:
Merge branch 'kaviharjani_12134' of github.com:KaviHarjani/twisted into kaviharjani_12134

KaviHarjani committed:
Fixing the text in code

KaviHarjani commented:
@adiroiban appreciate your reviews and apologise, I think I made most of the changes you requested except for the test one, not sure as you said custom handshake would be too much,  and that is why I marked those as resolved, I added comments with reference to the change

I have made changes and added an Exception, giving a better explanation of what is going on when cipher is wrong

Please let me know if it is good to go

Also, one more request @glyph @adiroiban if any one of you guys could help me with the patch, not sure why it says the code isn't covered and points to the uncovered test code

adiroiban commented on src/twisted/internet/test/test_endpoints.py:
```diff
@@ -3259,6 +3259,41 @@ def test_sslNoTrailingNewlinePem(self):
+                "certKey=%s:cipher=bla, blah"
+                % (escapedPEMPathName, escapedPEMPathName),
+            )
+            self.assertTrue("Invalid cipher list passed" in str(context.exception))
```
This line is never called as the previous lines raises an exception.

This is why is best to avoid using the context manager

```python
exception = self.assertRaises(
    SomeException,
    endpoints.serverFromString,
    reactor=reactor,
    ...
)
```

adiroiban replied:
It's best to folow TDD

Write the test first, run the test and see if fail... then fix the code.

If you write the test first and it passes from the first run, it means that the test is not valid.

https://docs.twisted.org/en/stable/development/test-standard.html

KaviHarjani replied:
[#e9308d](https://github.com/twisted/twisted/pull/12137/commits/6e30f9f83d118633e7be1c489d00073549c1b285)

adiroiban commented on src/twisted/internet/endpoints.py:
```diff
@@ -1466,6 +1473,13 @@ def _parseSSL(
     )
+    if cipher:
+        try:
+            cipherBytes = cipher.replace(",", ":").encode("ascii")
```
can this line throw an SSLError ?

KaviHarjani replied:
[#e9308d](https://github.com/twisted/twisted/pull/12137/commits/6e30f9f83d118633e7be1c489d00073549c1b285)

adiroiban commented on src/twisted/internet/endpoints.py:
```diff
@@ -1466,6 +1473,13 @@ def _parseSSL(
+            cipherBytes = cipher.replace(",", ":").encode("ascii")
+            cf.getContext().set_cipher_list(cipherBytes)
+        except SSLError:
+            raise Exception("Invalid cipher list passed")
```
I don't think that raising a generic exception is a good idea.

And the SSLError details are lost.

-------

I have checked the other parts of twisted, and I see that they just let the low-level SSLError bubble to the caller.

No need for special handling of the error.

Just make sure that we have a test for the error case.

-------

I have checked the Twisted coding standard, but I could not find any guidence for how is best to handle the exceptions

https://docs.twisted.org/en/stable/development/coding-standard.html


--------

But I think that converting an SSLError into a generic Exception is not a good idea.

KaviHarjani replied:
[#e9308d](https://github.com/twisted/twisted/pull/12137/commits/6e30f9f83d118633e7be1c489d00073549c1b285)

adiroiban commented on src/twisted/internet/endpoints.py:
```diff
@@ -1466,6 +1473,13 @@ def _parseSSL(
+    if cipher:
+        try:
+            cipherBytes = cipher.replace(",", ":").encode("ascii")
+            cf.getContext().set_cipher_list(cipherBytes)
```
I guess that this will work.

By understanding is that context factories might not cache the context.

This is why, we might need to change the API to something like this ... but this will break all existing context factory implementation.

So I guess that this is somehow ok for now.

We can refactor once someone nees to used this code with a custom context factory.

 
```
            cf.setOpenSSLCipherList(cipherStr)
```

If we delegate the cipher list setting to the context factory, then we no longer need the TLS handshake tests for endpoints... we will just rely on the context factory tests.

KaviHarjani replied:
@adiroiban what do you suggest?
Should we go ahead with making changes to the API ?

adiroiban commented on src/twisted/internet/test/test_endpoints.py:
```diff
@@ -3259,6 +3259,41 @@ def test_sslNoTrailingNewlinePem(self):
+    @skipIf(skipSSL, skipSSLReason)
+    def test_sslWithCustomCipher(self):
+        """
+        A cipher list is supported, using the OpenSSL format.
+        The colon (:) from the OpenSSL format is replaced with a comma (,).
+        Limitations to the test -> Not checked the ciphers returned.
```
I think that is best to add the full explanation/excuse of why we have a mediocre test here.

We have the discussion over the PR, but the information nees to be in the code.

```suggestion
        A cipher list is supported, using the OpenSSL format.
        The colon (:) from the OpenSSL format is replaced with a comma (,).
        
        This test does not provides much functionality coverage as there is no API to retrieve the current configured ciper list.
        Testing this functionality requires doing a TLS handshake.
```

adiroiban reviewed (changes requested):
Thanks for the changes.

I don't think that the new test is correct... the assertion is never called.

Also, I am not sure that converting SSLError to a generic Exception is a good idea.
I think that at this point is best to just raise the low-level SSLError.... as we don't have any other better Twisted specific exception.

I am still trying to find a good reason for not writing a proper test for the happy path.

I think that we should use iosim here and do a TLS handshake, to check that the code works as expections.

We use a ContextFactory here...but the cipher is set on a specific context and not an the context factory itself.

Later in the code execution, the ContextFactory might return a different context instance, and then the cipher list is not actually used in the TLS handshake.

KaviHarjani committed:
Raising SSL error except for generic one

KaviHarjani committed:
Update src/twisted/internet/test/test_endpoints.py

Co-authored-by: Adi Roiban <adiroiban@gmail.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

adiroiban commented:
Just listening for a port will not trigger any TLS handshake.

I think that you need to pass a real factory that uses a protocol that implements `IHandshakeListener`

Then make a client connection to trigger the handshake and check the negotiated protocol... or check the protocol negotiation failure.

The test can run over localhost with a randomly allocated port.
